src/deepagents/backends/protocol.py

```diff
@@ -1,218 +1,200 @@
-"""Protocol definition for pluggable memory backends.
+"""Type definitions for backend interface.
 
-This module defines the BackendProtocol that all backend implementations
+This module defines the unified Backend interface that all backend implementations
 must follow. Backends can store files in different locations (state, filesystem,
 database, etc.) and provide a uniform interface for file operations.
 """
 
-from typing import TYPE_CHECKING, Optional, Protocol, runtime_checkable
-from langgraph.types import Command
+from abc import ABC, abstractmethod
+from collections.abc import Awaitable, Callable
+from dataclasses import dataclass
+from typing import Any
+
 from langchain.tools import ToolRuntime
 
 
-@runtime_checkable
-class _BackendProtocol(Protocol):
-    """Protocol for pluggable memory backends.
+@dataclass
+class WriteResult:
+    """Result from backend write operations.
+
+    Attributes:
+        error: Error message on failure, None on success.
+        path: Absolute path of written file, None on failure.
+        files_update: State update dict for checkpoint backends, None for external storage.
+            Checkpoint backends populate this with {file_path: file_data} for LangGraph state.
+            External backends set None (already persisted to disk/S3/database/etc).
+
+    Examples:
+        >>> # Checkpoint storage
+        >>> WriteResult(path="/f.txt", files_update={"/f.txt": {...}})
+        >>> # External storage
+        >>> WriteResult(path="/f.txt", files_update=None)
+        >>> # Error
+        >>> WriteResult(error="File exists")
+    """
+
+    error: str | None = None
+    path: str | None = None
+    files_update: dict[str, Any] | None = None
+
+
+@dataclass
+class EditResult:
+    """Result from backend edit operations.
+
+    Attributes:
+        error: Error message on failure, None on success.
+        path: Absolute path of edited file, None on failure.
+        files_update: State update dict for checkpoint backends, None for external storage.
+            Checkpoint backends populate this with {file_path: file_data} for LangGraph state.
+            External backends set None (already persisted to disk/S3/database/etc).
+        occurrences: Number of replacements made, None on failure.
+
+    Examples:
+        >>> # Checkpoint storage
+        >>> EditResult(path="/f.txt", files_update={"/f.txt": {...}}, occurrences=1)
+        >>> # External storage
+        >>> EditResult(path="/f.txt", files_update=None, occurrences=2)
+        >>> # Error
+        >>> EditResult(error="File not found")
+    """
+
+    error: str | None = None
+    path: str | None = None
```

Comment:
We can either standardize what the backends return (e.g., path, occurrences) this allows us to standardize formatting also.

and/oR we allow the backends to ship a free form `msg`